File: kaji/ts/src/cli/list.ts
/**
 * `kaji list-integrations`: enumerate every integration in the registry
 * catalog (`registry/index.json`) and print its stable row contract:
 * `name  [tier]  version  description`.
 *
 * Follows the same flow as `kaji add` so the two commands agree on what
 * "available" means: an entry in `index.json`, not just a directory under
 * `registry/`.
 */
import type { RunOptions } from "@/cli/index";
import { TYPESCRIPT_SDK_CLI } from "@/cli/package-identity";
import {
  formatIntegrationError,
  loadManifest,
  loadRegistryIndex,
} from "@/integrations/registry-loader";

interface DiscoveryRow {
  name: string;
  version: string;
  stability: "experimental" | "beta";
  runtimes: string[];
  auth: { kind: "none" | "env" | "oauth"; provider: "google" | null };
  experimental_opt_in_required: boolean;
  next_commands: Record<string, string>;
}

function discoveryRow(manifest: Awaited<ReturnType<typeof loadManifest>>): DiscoveryRow {
  const experimental = manifest.stability === "experimental";
  const next_commands: Record<string, string> = {};
  for (const runtime of [...manifest.runtimes].sort()) {
    let command: string;
    if (manifest.auth.kind === "oauth") {
      command =
        runtime === "python"
          ? `python -m kaji.cli connect ${manifest.name} --principal <stable-host-principal-id>`
          : `${TYPESCRIPT_SDK_CLI} connect ${manifest.name} --principal <stable-host-principal-id>`;
    } else {
      command =
        runtime === "python"
          ? `python -m kaji.cli add ${manifest.name}`
          : `${TYPESCRIPT_SDK_CLI} add ${manifest.name}`;
      if (experimental) command += " --allow-experimental";
    }
    next_commands[runtime] = command;
  }
  return {
    name: manifest.name,
    version: manifest.version,
    stability: manifest.stability,
    runtimes: [...manifest.runtimes].sort(),
    auth: {
      kind: manifest.auth.kind,
      provider: manifest.auth.kind === "oauth" ? manifest.auth.provider : null,
    },
    experimental_opt_in_required: experimental,
    next_commands,
  };
}

export async function listIntegrations(rest: string[], opts: RunOptions): Promise<number> {
  const log = opts.log ?? ((m: string) => console.log(m));
  const err = opts.err ?? ((m: string) => console.error(m));
  const json = rest.length === 1 && rest[0] === "--json";
  if (rest.length !== 0 && !json) {
    err("usage: kaji list-integrations [--json]");
    return 2;
  }
  let index;
  try {
    index = await loadRegistryIndex(opts.registryRoot, { schemaRoot: opts.schemaRoot });
  } catch (error) {
    err(formatIntegrationError(error));
    return 1;
  }
  const entries = Object.keys(index.integrations).sort((a, b) => a.localeCompare(b));
  if (entries.length === 0 && !json) {
    log("No integrations found.");
    return 0;
  }
  const rows: DiscoveryRow[] = [];
  for (const name of entries) {
    try {
      const manifest = await loadManifest(opts.registryRoot, name, {
        schemaRoot: opts.schemaRoot,
        index,
      });
      rows.push(discoveryRow(manifest));
    } catch (error) {
      err(formatIntegrationError(error));
      return 1;
    }
  }
  if (json) {
    log(JSON.stringify(rows));
    return 0;
  }
  for (const row of rows) {
    const auth =
      row.auth.provider === null ? row.auth.kind : `${row.auth.kind}:${row.auth.provider}`;
    log(
      `${row.name}  [${row.stability}]  v${row.version}  auth=${auth}  runtimes=${row.runtimes.join(",")}`,
    );
    for (const [runtime, command] of Object.entries(row.next_commands)) {
      log(`  ${runtime}: ${command}`);
    }
  }
  return 0;
}

File: kaji/packages/py/src/cli/list.py
"""`kaji list-integrations` -- enumerate the registry."""

from __future__ import annotations

import argparse
import json as _json
from typing import TypedDict

from kaji.integrations import (
    Manifest,
    ManifestError,
    list_integrations as _list,
    load_manifest,
)

from ._pkg import TYPESCRIPT_SDK_CLI


def add_parser(sub: argparse._SubParsersAction) -> None:
    p = sub.add_parser(
        "list-integrations",
        help="list integrations available via `kaji add`",
    )
    p.add_argument("--json", action="store_true")
    p.set_defaults(func=run)


class _AuthRow(TypedDict):
    kind: str
    provider: str | None


class _IntegrationRow(TypedDict):
    name: str
    version: str
    stability: str
    runtimes: list[str]
    auth: _AuthRow
    experimental_opt_in_required: bool
    next_commands: dict[str, str]


def _next_commands(manifest: Manifest) -> dict[str, str]:
    experimental = manifest.stability == "experimental"
    commands: dict[str, str] = {}
    for runtime in sorted(manifest.runtimes):
        if manifest.auth.kind == "oauth":
            if runtime == "python":
                command = (
                    f"python -m kaji.cli connect {manifest.name} "
                    "--principal <stable-host-principal-id>"
                )
            else:
                command = (
                    f"{TYPESCRIPT_SDK_CLI} connect {manifest.name} "
                    "--principal <stable-host-principal-id>"
                )
        elif runtime == "python":
            command = f"python -m kaji.cli add {manifest.name}"
            if experimental:
                command += " --allow-experimental"
        else:
            command = f"{TYPESCRIPT_SDK_CLI} add {manifest.name}"
            if experimental:
                command += " --allow-experimental"
        commands[runtime] = command
    return commands


def _row(manifest: Manifest) -> _IntegrationRow:
    return {
        "name": manifest.name,
        "version": manifest.version,
        "stability": manifest.stability,
        "runtimes": sorted(manifest.runtimes),
        "auth": {
            "kind": manifest.auth.kind,
            "provider": manifest.auth.provider,
        },
        "experimental_opt_in_required": manifest.stability == "experimental",
        "next_commands": _next_commands(manifest),
    }


def run(args: argparse.Namespace) -> int:
    try:
        manifests = [load_manifest(name) for name in sorted(_list())]
    except ManifestError as error:
        print(f"Registry error: {error}")
        return 1
    rows = [_row(manifest) for manifest in manifests]
    if args.json:
        data = rows
        print(_json.dumps(data, indent=2))
        return 0

    if not manifests:
        print("No integrations available.")
        return 0

    for row in rows:
        auth = row["auth"]
        provider = auth["provider"]
        auth_label = auth["kind"] if provider is None else f"{auth['kind']}:{provider}"
        runtimes = ",".join(row["runtimes"])
        print(
            f"{row['name']}  [{row['stability']}]  v{row['version']}  "
            f"auth={auth_label}  runtimes={runtimes}"
        )
        commands = row["next_commands"]
        for runtime, command in commands.items():
            print(f"  {runtime}: {command}")
    return 0
